File: frontend/src/lib/services/emailAgentService.ts
import { apiClient } from "@/lib/api";

export type AgentConsentType =
  | "MAILBOX_ACCESS"
  | "CONTENT_CLASSIFICATION"
  | "DIGEST_NOTIFICATIONS"
  | "GOOGLE_CALENDAR"
  | "GOOGLE_TASKS"
  | "AI_PROCESSING";

export interface AgentConsent {
  consentType: AgentConsentType;
  granted: boolean;
  version: string;
  grantedAt?: string | null;
  revokedAt?: string | null;
}

export interface UserMailboxSummary {
  id: string;
  emailAddress: string;
  displayName?: string | null;
  provider: "GMAIL_OAUTH" | "IMAP_GENERIC";
  syncEnabled: boolean;
  lastSyncAt?: string | null;
  lastSyncStatus?: string | null;
  lastSyncError?: string | null;
  status: string;
  createdAt: string;
}

export interface TriageMessage {
  id: string;
  fromAddress: string;
  subject: string;
  snippet?: string | null;
  receivedAt: string;
  classification?: string | null;
  confidence?: string | null;
  reviewStatus: string;
  labels: string[];
  applicationId?: string | null;
  companyId?: string | null;
}

export interface ApplicationLinkSuggestion {
  applicationId: string;
  companyId: string | null;
  companyName: string | null;
  position: string;
  score: number;
}

export interface ProposedAgentActions {
  messageId: string;
  classification?: string | null;
  task:
    | { allowed: true; title: string; notes: string }
    | { allowed: false; reason: string };
  calendar:
    | { allowed?: true; kind: string; decision?: string; title?: string; reason?: string; message?: string }
    | { allowed: false; reason: string };
}

export interface AgentStatusResponse {
  success: boolean;
  agentEnabled: boolean;
  emailVerified: boolean;
  access: {
    allowed: boolean;
    reason: string;
    canConnectMailbox: boolean;
    canReadMailbox: boolean;
  };
  consentVersion: string;
  consentTypes: AgentConsentType[];
  consents: AgentConsent[];
  hasRequiredConsents: boolean;
  mailboxes: UserMailboxSummary[];
  pendingTriageCount: number;
}

export const CONSENT_LABELS: Record<AgentConsentType, { title: string; description: string }> = {
  MAILBOX_ACCESS: {
    title: "Accès aux boîtes mail",
    description:
      "Autoriser JobbingTrack à lire vos emails de recherche d'emploi (lecture seule, révocable).",
  },
  CONTENT_CLASSIFICATION: {
    title: "Classification automatique",
    description:
      "Analyser le contenu pour détecter refus, entretiens, relances et propositions.",
  },
  DIGEST_NOTIFICATIONS: {
    title: "Digest et notifications",
    description:
      "Recevoir un récap quotidien (18h) et des alertes utiles via JobbingTrack.",
  },
  GOOGLE_CALENDAR: {
    title: "Google Calendar",
    description:
      "Créer ou proposer des événements calendrier (sans horaire inventé).",
  },
  GOOGLE_TASKS: {
    title: "Google Tasks",
    description: "Synchroniser tâches et relances avec Google Tasks.",
  },
  AI_PROCESSING: {
    title: "Traitement IA (optionnel)",
    description: "Enrichissement assisté par IA locale — opt-in séparé, améliorable ensuite.",
  },
};

export async function fetchAgentStatus() {
  const { data } = await apiClient.get<AgentStatusResponse>("/email-agent/status");
  return data;
}

export async function updateAgentConsents(
  consents: Array<{ consentType: AgentConsentType; granted: boolean }>,
) {
  const { data } = await apiClient.put("/email-agent/consents", { consents });
  return data;
}

export async function startGoogleOAuth() {
  const { data } = await apiClient.get<{ success: boolean; authorizationUrl: string }>(
    "/email-agent/oauth/google/start",
  );
  return data;
}

export interface ImapDiscoveryResult {
  found: boolean;
  reason?: string;
  emailAddress?: string;
  domain?: string;
  suggested?: {
    imapHost: string;
    imapPort: number;
    imapUseTls: boolean;
    provider?: string;
    source?: string;
    note?: string | null;
    confidence?: string;
  };
  alternatives?: Array<{
    imapHost: string;
    imapPort: number;
    imapUseTls: boolean;
    provider?: string;
    source?: string;
  }>;
}

export async function discoverImapSettings(emailAddress: string) {
  const { data } = await apiClient.get<ImapDiscoveryResult & { success: boolean }>(
    "/email-agent/mailboxes/imap/discover",
    { params: { email: emailAddress } },
  );
  return data;
}

export async function connectImapMailbox(payload: {
  emailAddress: string;
  password: string;
  imapHost: string;
  imapPort?: number;
  imapUseTls?: boolean;
  smtpHost?: string;
  smtpPort?: number;
  displayName?: string;
}) {
  const { data } = await apiClient.post("/email-agent/mailboxes/imap", payload);
  return data;
}

export async function revokeMailbox(mailboxId: string) {
  const { data } = await apiClient.delete(`/email-agent/mailboxes/${mailboxId}`);
  return data;
}

export async function syncMailboxesNow() {
  const { data } = await apiClient.post("/email-agent/sync");
  return data;
}

export async function fetchTriageMessages(status = "PENDING") {
  const { data } = await apiClient.get<{ success: boolean; messages: TriageMessage[] }>(
    "/email-agent/triage",
    { params: { status } },
  );
  return data;
}

export async function reviewTriageMessage(
  messageId: string,
  reviewStatus: string,
  applicationId?: string,
) {
  const { data } = await apiClient.patch(`/email-agent/triage/${messageId}`, {
    reviewStatus,
    ...(applicationId ? { applicationId } : {}),
  });
  return data;
}

export async function fetchLinkSuggestions(messageId: string) {
  const { data } = await apiClient.get<{
    success: boolean;
    suggestions: ApplicationLinkSuggestion[];
  }>(`/email-agent/triage/${messageId}/link-suggestions`);
  return data;
}

export async function linkTriageToApplication(messageId: string, applicationId: string) {
  const { data } = await apiClient.post(`/email-agent/triage/${messageId}/link`, {
    applicationId,
  });
  return data;
}

export async function fetchProposedActions(messageId: string) {
  const { data } = await apiClient.get<{ success: boolean; actions: ProposedAgentActions }>(
    `/email-agent/triage/${messageId}/actions`,
  );
  return data;
}

export async function createTriageGoogleTask(messageId: string, payload?: { title?: string; notes?: string }) {
  const { data } = await apiClient.post(`/email-agent/triage/${messageId}/actions/task`, payload || {});
  return data;
}

export async function createTriageCalendarEvent(
  messageId: string,
  payload?: {
    hasExplicitTime?: boolean;
    hour?: number;
    minute?: number;
    startDateTime?: string;
    durationMinutes?: number;
    title?: string;
  },
) {
  const { data } = await apiClient.post(`/email-agent/triage/${messageId}/actions/calendar`, payload || {});
  return data;
}

export async function setUserAgentEnabled(userId: string, enabled: boolean) {
  const { data } = await apiClient.put(`/email-agent/users/${userId}/agent-enabled`, {
    enabled,
  });
  return data;
}

File: mobile/lib/services/email_agent_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:jobbingtrack_mobile/services/api_service.dart';
import 'package:jobbingtrack_mobile/services/http_correlation.dart';

class EmailAgentConsent {
  final String consentType;
  final bool granted;

  const EmailAgentConsent({required this.consentType, required this.granted});

  factory EmailAgentConsent.fromJson(Map<String, dynamic> json) {
    return EmailAgentConsent(
      consentType: json['consentType']?.toString() ?? '',
      granted: json['granted'] == true,
    );
  }
}

class EmailAgentMailbox {
  final String id;
  final String emailAddress;
  final String? displayName;
  final String provider;
  final bool syncEnabled;
  final String? lastSyncStatus;
  final DateTime? lastSyncAt;

  const EmailAgentMailbox({
    required this.id,
    required this.emailAddress,
    this.displayName,
    required this.provider,
    required this.syncEnabled,
    this.lastSyncStatus,
    this.lastSyncAt,
  });

  factory EmailAgentMailbox.fromJson(Map<String, dynamic> json) {
    return EmailAgentMailbox(
      id: json['id']?.toString() ?? '',
      emailAddress: json['emailAddress']?.toString() ?? '',
      displayName: json['displayName']?.toString(),
      provider: json['provider']?.toString() ?? '',
      syncEnabled: json['syncEnabled'] == true,
      lastSyncStatus: json['lastSyncStatus']?.toString(),
      lastSyncAt: DateTime.tryParse(json['lastSyncAt']?.toString() ?? ''),
    );
  }
}

class EmailAgentTriageMessage {
  final String id;
  final String fromAddress;
  final String subject;
  final String? snippet;
  final DateTime receivedAt;
  final String? classification;
  final String reviewStatus;
  final String? applicationId;

  const EmailAgentTriageMessage({
    required this.id,
    required this.fromAddress,
    required this.subject,
    this.snippet,
    required this.receivedAt,
    this.classification,
    required this.reviewStatus,
    this.applicationId,
  });

  factory EmailAgentTriageMessage.fromJson(Map<String, dynamic> json) {
    return EmailAgentTriageMessage(
      id: json['id']?.toString() ?? '',
      fromAddress: json['fromAddress']?.toString() ?? '',
      subject: json['subject']?.toString() ?? '(sans objet)',
      snippet: json['snippet']?.toString(),
      receivedAt: DateTime.tryParse(json['receivedAt']?.toString() ?? '') ?? DateTime.now(),
      classification: json['classification']?.toString(),
      reviewStatus: json['reviewStatus']?.toString() ?? 'PENDING',
      applicationId: json['applicationId']?.toString(),
    );
  }
}

class EmailAgentLinkSuggestion {
  final String applicationId;
  final String? companyName;
  final String position;
  final double score;

  const EmailAgentLinkSuggestion({
    required this.applicationId,
    this.companyName,
    required this.position,
    required this.score,
  });

  factory EmailAgentLinkSuggestion.fromJson(Map<String, dynamic> json) {
    return EmailAgentLinkSuggestion(
      applicationId: json['applicationId']?.toString() ?? '',
      companyName: json['companyName']?.toString(),
      position: json['position']?.toString() ?? '',
      score: (json['score'] as num?)?.toDouble() ?? 0,
    );
  }
}

class EmailAgentStatus {
  final bool agentEnabled;
  final bool hasRequiredConsents;
  final bool accessAllowed;
  final String accessReason;
  final List<EmailAgentConsent> consents;
  final List<EmailAgentMailbox> mailboxes;
  final int pendingTriageCount;

  const EmailAgentStatus({
    required this.agentEnabled,
    required this.hasRequiredConsents,
    required this.accessAllowed,
    required this.accessReason,
    required this.consents,
    required this.mailboxes,
    required this.pendingTriageCount,
  });

  factory EmailAgentStatus.fromJson(Map<String, dynamic> json) {
    final access = json['access'] as Map<String, dynamic>? ?? {};
    final consentsRaw = json['consents'] as List<dynamic>? ?? [];
    final mailboxesRaw = json['mailboxes'] as List<dynamic>? ?? [];
    return EmailAgentStatus(
      agentEnabled: json['agentEnabled'] == true,
      hasRequiredConsents: json['hasRequiredConsents'] == true,
      accessAllowed: access['allowed'] == true,
      accessReason: access['reason']?.toString() ?? '',
      consents: consentsRaw
          .map((e) => EmailAgentConsent.fromJson(Map<String, dynamic>.from(e as Map)))
          .toList(),
      mailboxes: mailboxesRaw
          .map((e) => EmailAgentMailbox.fromJson(Map<String, dynamic>.from(e as Map)))
          .toList(),
      pendingTriageCount: (json['pendingTriageCount'] as num?)?.toInt() ?? 0,
    );
  }
}

class ImapDiscoverySuggestion {
  final String imapHost;
  final int imapPort;
  final bool imapUseTls;
  final String? provider;
  final String? source;
  final String? note;

  const ImapDiscoverySuggestion({
    required this.imapHost,
    required this.imapPort,
    required this.imapUseTls,
    this.provider,
    this.source,
    this.note,
  });

  factory ImapDiscoverySuggestion.fromJson(Map<String, dynamic> json) {
    return ImapDiscoverySuggestion(
      imapHost: json['imapHost']?.toString() ?? '',
      imapPort: (json['imapPort'] as num?)?.toInt() ?? 993,
      imapUseTls: json['imapUseTls'] != false,
      provider: json['provider']?.toString(),
      source: json['source']?.toString(),
      note: json['note']?.toString(),
    );
  }
}

class EmailAgentService {
  static const Duration _imapTimeout = Duration(seconds: 90);
  static const Duration _defaultTimeout = Duration(seconds: 30);

  static Map<String, String> _headers(String? token) =>
      HttpCorrelation.jsonHeaders(bearerToken: token);

  static Future<Map<String, dynamic>> _decode(http.Response response, String action) async {
    Map<String, dynamic> data = {};
    try {
      final parsed = jsonDecode(response.body);
      if (parsed is Map<String, dynamic>) data = parsed;
    } catch (_) {}
    if (response.statusCode >= 200 && response.statusCode < 300) {
      return data;
    }
    final msg = data['message']?.toString() ??
        data['error']?.toString() ??
        'Erreur $action (HTTP ${response.statusCode})';
    throw Exception(msg);
  }

  static Future<EmailAgentStatus> fetchStatus({required String? token}) async {
    final res = await http
        .get(
          Uri.parse('${ApiService.baseUrl}/api/v1/email-agent/status'),
          headers: _headers(token),
        )
        .timeout(_defaultTimeout);
    final data = await _decode(res, 'statut agent');
    return EmailAgentStatus.fromJson(data);
  }

  static Future<void> updateConsents({
    required String? token,
    required List<Map<String, dynamic>> consents,
  }) async {
    final res = await http
        .put(
          Uri.parse('${ApiService.baseUrl}/api/v1/email-agent/consents'),
          headers: _headers(token),
          body: jsonEncode({'consents': consents}),
        )
        .timeout(_defaultTimeout);
    await _decode(res, 'consentements');
  }

  static Future<ImapDiscoverySuggestion?> discoverImap({
    required String? token,
    required String emailAddress,
  }) async {
    final uri = Uri.parse('${ApiService.baseUrl}/api/v1/email-agent/mailboxes/imap/discover')
        .replace(queryParameters: {'email': emailAddress});
    final res = await http.get(uri, headers: _headers(token)).timeout(_defaultTimeout);
    final data = await _decode(res, 'découverte IMAP');
    if (data['found'] != true) return null;
    final suggested = data['suggested'];
    if (suggested is! Map) return null;
    return ImapDiscoverySuggestion.fromJson(Map<String, dynamic>.from(suggested));
  }

  static Future<EmailAgentMailbox> connectImap({
    required String? token,
    required String emailAddress,
    required String password,
    required String imapHost,
    int imapPort = 993,
    bool imapUseTls = true,
    String? displayName,
  }) async {
    final res = await http
        .post(
          Uri.parse('${ApiService.baseUrl}/api/v1/email-agent/mailboxes/imap'),
          headers: _headers(token),
          body: jsonEncode({
            'emailAddress': emailAddress,
            'password': password,
            'imapHost': imapHost,
            'imapPort': imapPort,
            'imapUseTls': imapUseTls,
            if (displayName != null && displayName.isNotEmpty) 'displayName': displayName,
          }),
        )
        .timeout(_imapTimeout);
    final data = await _decode(res, 'connexion IMAP');
    final mailbox = data['mailbox'];
    if (mailbox is! Map) throw Exception('Réponse boîte mail invalide');
    return EmailAgentMailbox.fromJson(Map<String, dynamic>.from(mailbox));
  }

  static Future<void> revokeMailbox({required String? token, required String mailboxId}) async {
    final res = await http
        .delete(
          Uri.parse('${ApiService.baseUrl}/api/v1/email-agent/mailboxes/$mailboxId'),
          headers: _headers(token),
        )
        .timeout(_defaultTimeout);
    await _decode(res, 'révocation boîte');
  }

  static Future<List<Map<String, dynamic>>> syncNow({required String? token}) async {
    final res = await http
        .post(
          Uri.parse('${ApiService.baseUrl}/api/v1/email-agent/sync'),
          headers: _headers(token),
        )
        .timeout(_imapTimeout);
    final data = await _decode(res, 'synchronisation');
    final results = data['results'];
    if (results is List) {
      return results.map((e) => Map<String, dynamic>.from(e as Map)).toList();
    }
    return [];
  }

  static Future<List<EmailAgentTriageMessage>> fetchTriage({
    required String? token,
    String status = 'PENDING',
  }) async {
    final uri = Uri.parse('${ApiService.baseUrl}/api/v1/email-agent/triage')
        .replace(queryParameters: {'status': status});
    final res = await http.get(uri, headers: _headers(token)).timeout(_defaultTimeout);
    final data = await _decode(res, 'triage');
    final messages = data['messages'] as List<dynamic>? ?? [];
    return messages
        .map((e) => EmailAgentTriageMessage.fromJson(Map<String, dynamic>.from(e as Map)))
        .toList();
  }

  static Future<void> reviewTriage({
    required String? token,
    required String messageId,
    required String reviewStatus,
  }) async {
    final res = await http
        .patch(
          Uri.parse('${ApiService.baseUrl}/api/v1/email-agent/triage/$messageId'),
          headers: _headers(token),
          body: jsonEncode({'reviewStatus': reviewStatus}),
        )
        .timeout(_defaultTimeout);
    await _decode(res, 'revue triage');
  }

  static Future<List<EmailAgentLinkSuggestion>> fetchLinkSuggestions({
    required String? token,
    required String messageId,
  }) async {
    final res = await http.get(
      Uri.parse('${ApiService.baseUrl}/api/v1/email-agent/triage/$messageId/link-suggestions'),
      headers: _headers(token),
    ).timeout(_defaultTimeout);
    final data = await _decode(res, 'suggestions liaison');
    final raw = data['suggestions'] as List<dynamic>? ?? [];
    return raw
        .map((e) => EmailAgentLinkSuggestion.fromJson(Map<String, dynamic>.from(e as Map)))
        .toList();
  }

  static Future<void> linkToApplication({
    required String? token,
    required String messageId,
    required String applicationId,
  }) async {
    final res = await http.post(
      Uri.parse('${ApiService.baseUrl}/api/v1/email-agent/triage/$messageId/link'),
      headers: _headers(token),
      body: jsonEncode({'applicationId': applicationId}),
    ).timeout(_defaultTimeout);
    await _decode(res, 'liaison candidature');
  }
}

const Map<String, String> emailAgentConsentLabels = {
  'MAILBOX_ACCESS': 'Accès aux boîtes mail',
  'CONTENT_CLASSIFICATION': 'Classification automatique',
  'DIGEST_NOTIFICATIONS': 'Digest et notifications',
  'GOOGLE_CALENDAR': 'Google Calendar',
  'GOOGLE_TASKS': 'Google Tasks',
  'AI_PROCESSING': 'Traitement IA (optionnel)',
};

const List<String> emailAgentConsentOrder = [
  'MAILBOX_ACCESS',
  'CONTENT_CLASSIFICATION',
  'DIGEST_NOTIFICATIONS',
  'GOOGLE_CALENDAR',
  'GOOGLE_TASKS',
  'AI_PROCESSING',
];
